pkg/platform/authentication/auth.go

```diff
@@ -215,6 +249,57 @@ func (s *Auth) AuthenticateWithToken(token string) error {
 	})
 }
 
+// AuthenticateWithDeviceCode posts a request to authenticate this device on the Platform and waits
+// for the user to authorize the request.
+// The given callback function is called when the user should be prompted to authorize the request.
+// Returns an error if authentication cannot be performed (e.g. timeout or Platform is unreachable).
+func (s *Auth) AuthenticateWithDevice(promptCallback func(userCode, uri string)) error {
+	if s.Authenticated() {
+		return nil // nothing to do
+	}
+	// Post the authentication request to the Platform.
+	postParams := oauth.NewAuthDevicePostParams()
+	response, err := mono.Get().Oauth.AuthDevicePost(postParams)
+	if err != nil {
+		logging.Error("Error requesting device authentication: %v", err)
+		return locale.NewError("err_auth_device")
+	}
+	// Prompt the user to authorize the request.
+	promptCallback(*response.Payload.UserCode, *response.Payload.VerificationURIComplete)
+	// Wait for the authorization.
+	deviceCode := strfmt.UUID(*response.Payload.DeviceCode)
+	getParams := oauth.NewAuthDeviceGetParams()
+	getParams.SetDeviceCode(deviceCode)
+	startTime := time.Now()
+	const timeout = 6 * 60 * time.Second
+	for {
+		response, err := mono.Get().Oauth.AuthDeviceGet(getParams)
+		if response != nil {
+			s.storeAuthenticatedDevice(deviceCode, response)
+			break
+		} else if errs.Matches(err, &oauth.AuthDeviceGetBadRequest{}) {
+			badRequest := err.(*oauth.AuthDeviceGetBadRequest)
+			errorString := *badRequest.Payload.Error
+			const expiredToken = "expired_token"
+			const invalidClient = "invalid_client"
+			const slowDown = "slow_down"
+			if errorString == expiredToken || time.Since(startTime) >= timeout {
+				return locale.NewInputError("auth_device_timeout")
+			} else if errorString == invalidClient {
+				logging.Error("Error requesting device authentication: invalid client") // IP address mismatch
+				return locale.NewError("err_auth_device")
+			} else if errorString == slowDown {
+				logging.Warning("Attempting to check for authorization status too frequently.")
+			}
+			time.Sleep(5 * time.Second) // then try again
```

Comment:
Yes, this is the polling. Carey told me that we need to stop polling on "expired_token" and "invalid_client". Other than that, poll away.

Reply:
That's missing one of my AC's though, which Carey hasn't been exposed to:

```
- We poll for the login success (ie. we do not show the ENTER prompt like the script Carey wrote)

  - BUT; when we first request the authorization via https://platform.activestate.com/api/v1/oauth/authorize/device this should optionally return a token which can tell the State tool NOT to pull and instead show the ENTER prompt

  - This is in case the polling turns out to create too high a load for our API/DB

  - This token is not implemented on the server at this time, nor is it meant to be. It is something we should be able to change API side after the fact if the need arises, but the State Tool already needs to be aware of this.
```

So the idea here is that we can at a later date update the API to give some sort of signal that it should not poll automatically and instead the user should press "Enter" to continue (ie. poll on Enter).

It doesn't feel like `expired_token` or `invalid_client` addresses this use-case, nor that we have the "Enter" prompt.

Reply:
I'm not sure what problem you're trying to solve @Naatan that the API doesn't solve already.  If clients poll faster than every 5 second they get a slow down warning which is why Mitchell has the 5 second wait here.  Why is a prompt necessary from the API?  You suggest deviate from the IETF spec which I'm going to push back on unless there is a really good reason to do so.

Reply:
This is a failsafe to save us from problems like the one we had recently where we were sending billions of events, if that ever happened on our own API's we'd be swamped. I discussed this with Shane and we felt this would be a cheap solution. To be clear no action is required on your end @th3coop other than to advice if there is a sensible place in the payload that could carry this signal to the client.

So to be clear; if all goes well then we would never use this and the API would never have to support it. But IF we end up having to need it because the API is being overwhelmed this is what would happen:

- The client calls the API as normal to start the auth flow, but it has a simple check client-side that validates that if the API responds with some additional parameter then we should not poll and instead prompt the user to hit ENTER.
- We deploy a change to the API that will now have it provide this parameter
- The client sees this and prompts the user to press ENTER and saves us from having our API hammered
- Crisis averted.

This ought to be a really simple tweak we can make client-side that will allow us to avoid a big crisis. Low cost, high value. We can come up with a parameter of our own, but I figure it would be better for you to provide one so we're not making any false assumptions.

Reply:
I don't understand how this does anything better than what the `slow_down` warning does.  The `slow_down` warning will prevent exactly the scenario you describe but your suggest will require not only the API to change but all 3rd party clients to account for and use the new API data.  Why deviate from how the DCAG spec defined to handle this scenario?  Just seems like you're adding complication with no benegit.